daphne/testing.py

```diff
@@ -167,19 +172,28 @@ class TestApplication:
     setup_storage = os.path.join(tempfile.gettempdir(), "setup.testio")
     result_storage = os.path.join(tempfile.gettempdir(), "result.testio")
 
-    def __init__(self, scope):
+    def __init__(self, scope, lock):
         self.scope = scope
+        self.lock = lock
         self.messages = []
 
     async def __call__(self, send, receive):
         # Receive input and send output
         logging.debug("test app coroutine alive")
+
+        async def receive_and_acquire_lock():
+            received = await receive()
+            self.lock.acquire()
```

Comment:
Still small window for the same race here. I guess what is desired is to acquire the lock at the beginning or somewhere in the body of `receive()`. Will take a look if it can be done cleanly

Reply:
What is needed here is ability to acquire the lock right when the coroutine starts; this can't be expressed in a wrapper async function like this since it needs to happen "inside" the await. Maybe there is some utility for this...

Reply:
Not sure my last comment even makes sense, but in any case I've tested this a bit by inserting an async sleep before the lock acquisition. For whatever reason I can push it up to about 1 second exactly before it induces the failure again.

As for actually plumbing in the lock acquisition to occur just prior to `receive` actually doing this work, I think this would require overloading the `_get` method on the `asyncio.Queue` used by `daphne.server.Server`. This could be done in some way where it is not necessarily polluting the server with test-rig code, but rather providing some generic hook. But it feels ugly to me and also like overkill.

I will poke around a bit more and maybe try to understand the bigger picture a bit, but feel free to chime in, anybody

Reply:
The 1 second is not inducing the race-condition; only a timeout. Maybe this is "safe" practically speaking... good enough for test code, maybe